### ISSUE TITLE
Ceph-CSI Driver support for multiple ceph clusters from a single operator

### ISSUE THREAD

**Issue by travisn** [rook/rook#3331]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
The ceph-csi integration with rook currently requires a single ceph cluster to be configured. This is due to the mon endpoints being updated on a per-cluster basis as seen in #3271. This will work fine as long as there is only ceph cluster deployed per operator. 

To support multiple clusters per operator, we would need the following approach:
- The operator will deploy the CSI driver:
   - An empty configmap will be created in the same namespace. The CSI driver needs to mount this in order to launch
   - The CSI driver is deployed in the same namespace as the operator
- Every time the operator creates/updates the mons in a cluster, it would do the following:
   - Take a lock (There are multiple goroutines that update mons, one for each cluster's mon health check and one for the reconcile loop)
   - Update a configmap in the same namespace as the operator (and csi driver). The configmap follows the format expected by the csi driver and would contain a list of all the clusters with all their mon endpoints.
   - The output of [`FormatCsiClusterConfig`](https://github.com/phlogistonjohn/rook/blob/master/pkg/operator/ceph/cluster/mon/mon.go#L469) will be written to the new configmap rather than `rook-ceph-mon-endpoints`. 
   - Release the lock

The lock (`sync.Mutex`) can be defined by the [`ClusterController`](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/controller.go#L93) since there is only one cluster controller defined for the operator. When the cluster controller creates a new cluster, it would pass the needed context* to [`newCluster()`](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/controller.go#L252). The context* would be passed onto the mon's [`Cluster`](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/cluster.go#L82) type for later use by the [`saveMonConfig()`](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/mon/mon.go#L455) method.

*The context from the previous paragraph is referring possibly to the `sync.Mutex` instance that can be used when the mons are updated. Alternatively, a callback method could be passed so that the details of locking around the mutex could be hidden from the mons. Either way should be fine in this case.

@ShyamsundarR @humblec @Madhu-1 @phlogistonjohn

**Comment by humblec** [comment]
@travisn thanks for writing this down. It matches with the thought or solution I have in mind. 

>The ceph-csi integration with rook currently requires a single ceph cluster to be configured. This is due to the mon endpoints being updated on a per-cluster basis as seen in #3271. This will work fine as long as there is only ceph cluster deployed per operator.
>
I doubt this work seemlessly in all the setup atm.

**Comment by Madhu-1** [comment]
> @travisn thanks for writing this down. It matches with the thought or solution I have in mind.
> 
> > The ceph-csi integration with rook currently requires a single ceph cluster to be configured. This is due to the mon endpoints being updated on a per-cluster basis as seen in #3271. This will work fine as long as there is only ceph cluster deployed per operator.
> 
> I doubt this work seemlessly in all the setup atm.

@humblec it will work if both operator and cluster are created in the same namespace

**Comment by Madhu-1** [comment]
just a note there will be one CSI per operator. if there are `N operator` each operator is managing  different clusters then there will be `N CSI pods`  deployed in operator namespaces

**Comment by humblec** [comment]
>@humblec it will work if both operator and cluster are created in the same namespace
Thats what my point was , is it always guaranteed that ceph cluster and operator are created in same namespace ?

**Comment by humblec** [comment]
> just a note there will be one CSI per operator. if there are `N operator` each operator is managing different clusters then there will be `N CSI pods` deployed in operator namespaces

That assumption is correct, but is there a scenario of `N operators`  ?

**Comment by Madhu-1** [comment]
> > @humblec it will work if both operator and cluster are created in the same namespace
> 
> Thats what my point was , is it always guaranteed that ceph cluster and operator are created in same namespace ?

its configurable https://github.com/rook/rook/blob/818ad9c7ca4d352a9bed202e619d47b05f1a4dab/cluster/examples/kubernetes/ceph/operator.yaml#L16-L18

https://github.com/rook/rook/blob/818ad9c7ca4d352a9bed202e619d47b05f1a4dab/cluster/examples/kubernetes/ceph/cluster.yaml#L13-L16

**Comment by humblec** [comment]
@Madhu-1 I am aware of that configuration :), the point I am making here is, we have to convey that to users ( may be via documentation) , failed to see it in the documentation. May be I missed, if not, I can send a PR

**Comment by ShyamsundarR** [comment]
Is there a case where there can be multiple operators? The above talks about a single operator managing multiple clusters, hence asking if the multiple operator scenario is also a possibility.

If yes, then the CSI drivername should be made specific to each operator name, such that multiple such instance can exist in a single Kubernetes cluster.

Other than the above detail, rest of the design/approach looks fine to me.

**Comment by travisn** [comment]
While not common, we should allow for multiple operators. Each operator would start in its own namespace and each operator would start its CSI driver in the same namespace as that operator. Where would the CSI driver name be visible? In the storage class?

**Comment by ShyamsundarR** [comment]
> While not common, we should allow for multiple operators. Each operator would start in its own namespace and each operator would start its CSI driver in the same namespace as that operator. Where would the CSI driver name be visible? In the storage class?

The drivername [1] would be used in the storage class, as the `provisioner` [2] field.

[1] drivername configuration:
cephfs: https://github.com/ceph/ceph-csi/blob/master/docs/deploy-cephfs.md#configuration
rbd: https://github.com/ceph/ceph-csi/blob/master/docs/deploy-rbd.md#configuration

[2] Use of drivername in the StorageClass:
cephfs (similarly in rbd): https://github.com/ceph/ceph-csi/blob/7f993707ca9f575fabd78926806d3a6532d3e356/examples/cephfs/storageclass.yaml#L1-L7

**Comment by jarrpa** [comment]
Sorry if I'm coming in blind, but the driver name shoul dbe configurable form Rook in this case. If multiple CSI drivers share the same driver name, they will all listen for the same provisioning events and it will be nondeterministic which one will pick up a given event first.

**Comment by ShyamsundarR** [comment]
> Sorry if I'm coming in blind, but the driver name shoul dbe configurable form Rook in this case. If multiple CSI drivers share the same driver name, they will all listen for the same provisioning events and it will be nondeterministic which one will pick up a given event first.

Yes, Rook SHOULD specialize the drivername as per the operator instance.

The latter half of the observation is right, we cannot have multiple CSI instances advertising the same name, it would just not work.

**Comment by julienlau** [comment]
This feature would be very helpful with rbd-mirror.
As is it is very difficult to failover from one cluster to another.
It would be nice if the csi driver plugged to the second cluster would be ready in standby.

I did not tested it yet, but maybe using the mounter : rbd-nbd (alpha feature) would enable ? Since this mounter acts in user space and not in kernel space it should be OK to pop multiple dirvers if all storage class use the rbd-nbd mounter ?

**Comment by travisn** [comment]
@julienlau This is supported since resolved over five years ago. What is missing? A new issue may be helpful to describe the full scenario. I assume you're also following the [rbd mirroring documentation](https://rook.io/docs/rook/latest-release/Storage-Configuration/Block-Storage-RBD/rbd-mirroring/).

**Comment by julienlau** [comment]
I don't understand your answer, so maybe I did not explain correctly my problem...
The RBD mirroring is supported, ok we agree.

Please note that I use external ceph clusters ! Rook does not manage ceph and ceph is hosted on premises outside of kubernetes.

Let's say I want to perform a planned failover.
The procedure would be : 
- Stop all clients that use the primary image. 
On kubernetes be aware to use reclaimPolicy=Retain on PVC.
- Demote primary image 
`rbd mirror image demote ${mypool}/${myimage}`
- Marks the image as non-primary (unwriteable to standard RBD clients)
- Creates a demote snapshot
- Promote non-primary image 
`rbd mirror image promote ${mypool}/${myimage}`
- Works only after demote snapshot is fully synced
- Creates a promote snapshot
- Marks the image as primary (writeable to standard RBD clients)
- After some time, check the status of the images from a monitor node in the site-b cluster. They should show a state of `up+stopped`.
- Stop the primary ceph cluster.
- **Resume the access to the images.**
On kubernetes this last stage is nightmarish ! 
1/ edit csi-rbd config to update clusterID and secrets to goes to the new clusterID !
2/ manually recreate all PVC and PV.

When the ceph cluster is down, csi-rbd cannot rely on the original clusterID to connect to ceph.
This would be less difficult, if the csi-rbd driver could directly be connected to both ceph clusters.

also see : https://rook.io/docs/rook/latest-release/Storage-Configuration/Block-Storage-RBD/rbd-async-disaster-recovery-failover-failback/#relocation

**Comment by Madhu-1** [comment]
@julienlau you can use the csi-addons to automate all the steps from 9. you can make use of https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/clusterid-mapping.md to create a mapping to point to the right clusterID. 

You can use velero or any backup tools to recreate all the PVC's and PV's? for example https://github.com/RamenDR/ramen does all the automation you have mentioned above.

**Comment by julienlau** [comment]
Thank you for the very relevant links. It seems that the solution is not simple, but at least I was able to explain my problem correctly!
I'll have an in-depth look to Ramen !

**Comment by julienlau** [comment]
It seems that cluster-mapping is convenient for that, but the doc is really obscure 
https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/clusterid-mapping.md
https://github.com/ceph/ceph-csi/issues/4493
https://github.com/ceph/ceph-csi/pull/4501
